src/cgutils.cpp

```diff
@@ -1601,10 +1601,18 @@ static void undef_var_error_ifnot(jl_codectx_t &ctx, Value *ok, jl_sym_t *name,
     ctx.builder.SetInsertPoint(ifok);
 }
 
+// ctx.builder.CreateIsNotNull(v) lowers incorrectly in non-standard
+// address spaces where null is not zero
+// TODO: adapt to https://github.com/llvm/llvm-project/pull/131557 once merged
 static Value *null_pointer_cmp(jl_codectx_t &ctx, Value *v)
 {
     ++EmittedNullchecks;
-    return ctx.builder.CreateIsNotNull(v);
+    Type *T = v->getType();
+    return ctx.builder.CreateICmpNE(
+            v,
+            ctx.builder.CreateAddrSpaceCast(
+                Constant::getNullValue(ctx.builder.getPtrTy(AddressSpace::Derived)),
+                PointerType::get(T, T->getPointerAddressSpace())));
```

Comment:
```suggestion
                Constant::getNullValue(ctx.builder.getPtrTy(0)), T);
```

I expect AMD to still have a very bad time of this, since LLVM does currently define these as being identical and expect it to constant fold away.

Reply:
I think they've forbidden this somewhat (this causes problems elsewhere but we should just follow what clang does for now). And yeah it should be addrspace 0, not derived here.